### PR TITLE
gnrc_pkt: introduce packet list operations

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
Up until now `utlist.h` is used to manipulate the packet list within a GNRC pkt snip. Since I always wanted to see if there would be an advantage when using `core`'s `list.h`, I provide this `gnrc_pktsnip_t`-specific API to manipulate the packet list. It has the additional advantage of type safety and that future steps could encapsulate those operations.

### Testing procedure
Read carefully, the changes should be obvious.

~~Binary size should remain unchanged.~~ (see https://github.com/RIOT-OS/RIOT/pull/15212#issuecomment-707800890)

None of the usages of `LL_...` remain in tree for manipulation of `gnrc_pktsnip_t` lists (**attention**: there _are_ other usages!).

All GNRC tests should still pass, `gnrc_networking`, `gnrc_networking_mac`, and `gnrc_border_router` examples should still work~~(but this should also be confirmable with the binaries remaining unchanged)~~.

### Issues/PRs references
None, but together with https://github.com/RIOT-OS/RIOT/pull/15211, there should be no usage of `utlist.h` macros remaining in the tree to manipulate GNRC packets.

### CODE REVIEW

File: tests/gnrc_sock_ip/stack.c [hunk 1]
```diff
@@ -62,13 +62,13 @@ static gnrc_pktsnip_t *_build_ipv6_packet(const ipv6_addr_t *src,
...
     ipv6_hdr->hl = 64;
-    LL_APPEND(payload, ipv6);
+    payload = gnrc_pkt_append(payload, ipv6);
     netif_hdr = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
     if (netif_hdr == NULL) {
         return NULL;
     }
     ((gnrc_netif_hdr_t *)netif_hdr->data)->if_pid = (kernel_pid_t)netif;
-    LL_APPEND(payload, netif_hdr);
+    payload = gnrc_pkt_append(payload, netif_hdr);
```
Comment (maribu) [inline tests/gnrc_sock_ip/stack.c:71]: How about
```suggestion
    return gnrc_pkt_append(payload, netif_hdr);
```
?

(And the `return payload;` below could be dropped. I'm somehow to stupid to include non-changed lines into the suggestion.)

File: tests/gnrc_sock_udp/stack.c [hunk 2]
```diff
@@ -85,13 +85,13 @@ static gnrc_pktsnip_t *_build_udp_packet(const ipv6_addr_t *src,
...
     }
-    LL_APPEND(udp, ipv6);
+    udp = gnrc_pkt_append(udp, ipv6);
     netif_hdr = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
     if (netif_hdr == NULL) {
         return NULL;
     }
     ((gnrc_netif_hdr_t *)netif_hdr->data)->if_pid = (kernel_pid_t)netif;
-    LL_APPEND(udp, netif_hdr);
+    udp = gnrc_pkt_append(udp, netif_hdr);
```
Comment (maribu) [inline tests/gnrc_sock_udp/stack.c:94]: Same here. (Directly return.)

**Review by maribu — COMMENTED** [review]
Code looks good to me. Have you tested that this doesn't change the binaries? If so, I'd ACK right away.

**Comment by miri64** [comment]
> Code looks good to me. Have you tested that this doesn't change the binaries?

Sample-wise, but not completely. Maybe @bergzand's tool can help here?

File: tests/gnrc_sock_ip/stack.c (hunk 1 above)
Comment (miri64) [reply tests/gnrc_sock_ip/stack.c]: Done.

File: tests/gnrc_sock_udp/stack.c (hunk 2 above)
Comment (miri64) [reply tests/gnrc_sock_udp/stack.c]: Done.

**Comment by miri64** [comment]
Nice catch, @maribu. Applied your suggestions.

**Comment by miri64** [comment]
One could use this script

```sh
#!/usr/bin/env python

import json
import urllib.request

MASTER = "https://ci.riot-os.org/RIOT-OS/RIOT/master/a83927d0313caf1ea2d98e50ad959a0bc2ef5408/sizes.json"
PR15212 = "https://ci.riot-os.org/RIOT-OS/RIOT/15212/1886a4867b3d162afcd948ae6c8a3f62c6d9426b/sizes.json"

def get_json(url):
    with urllib.request.urlopen(url) as f:
        assert f.status == 200
        return json.load(f)

master = get_json(MASTER)
pr15212 = get_json(PR15212)

assert master["sizes"].keys() == pr15212["sizes"].keys()
for app in master["sizes"]:
    if master["sizes"][app].keys() != pr15212["sizes"][app].keys():
        continue
    for board in master["sizes"][app]:
        if (master["sizes"][app][board]["text"] != pr15212["sizes"][app][board]["text"] or
            master["sizes"][app][board]["data"] != pr15212["sizes"][app][board]["data"] or
            master["sizes"][app][board]["bss"] != pr15212["sizes"][app][board]["bss"]):
            if board.startswith("esp"):
                continue
            print("{}[{}]:".format(app, board))
            print("          text     data      bss      dec")
            print("master  {text:8d} {data:8d} {bss:8d} {dec:8d}"
                  .format(**master["sizes"][app][board]))
            print("PR15212 {text:8d} {data:8d} {bss:8d} {dec:8d}"
                  .format(**pr15212["sizes"][app][board]))
```

To get the differences from Murdock. However, currently there is a lot of diff since a83927d (the commit this PR is currently based on) is quite old and 1886a48 includes current master. Let me rebase to the current nightly and see what that yields.

**Comment by miri64** [comment]
Rebased and squashed.

**Comment by miri64** [comment]
There will most likely be some diff in `tests/gnrc_tcp`, as #15211 was merged a moment ago.

**Comment by aabadie** [comment]
> There will most likely be some diff in tests/gnrc_tcp, as #15211 was merged a moment ago.

To compare with, you can also checkout `HEAD~2` starting from the branch of this PR.

**Comment by miri64** [comment]
> > There will most likely be some diff in tests/gnrc_tcp, as #15211 was merged a moment ago.
> 
> To compare with, you can also checkout `HEAD~2` starting from the branch of this PR.

There is no murdock build for `HEAD~2` == d92f280add25ca6e1bd70a71c365c8cd43919004 and I don't want to use up my PC's CPU for a day to build all apps ;-).

**Comment by aabadie** [comment]
> There is no murdock build for HEAD~2

Ah yes, I didn't notice the script was fetching the json output from Murdock. But you can still rebase the branch of this PR on top of the latest master build of Murdock. No need to use your local PC.

**Comment by miri64** [comment]
> Ah yes, I didn't notice the script was fetching the json output from Murdock. But you can still rebase the branch of this PR on top of the latest master build of Murdock. No need to use your local PC.

That's what I just did ;-) but the latest master build was [10 hours ago](https://ci.riot-os.org/RIOT-OS/RIOT/master/2df92b0d6165dda4095c99021ef58610361e45df/output.html). Since then already some stuff got merged into master which will be included in the build of this PR ;-)

**Comment by aabadie** [comment]
> Since then already some stuff got merged into master which will be included in the build of this PR ;-)

Yeah and what's the problem if you force push and wait 45 minutes to get the new results (this PR rebased on the 10h old version of master) ?

**Comment by miri64** [comment]
> > Since then already some stuff got merged into master which will be included in the build of this PR ;-)
> 
> Yeah and what's the problem if you force push and wait 45 minutes to get the new results (this PR rebased on the 10h old version of master) ?

That's what I did [21 minutes ago](https://github.com/RIOT-OS/RIOT/pull/15212#event-3870790500)... There is no problem, just that the Murdock build of this PR will include the _current_ master, because the PR is rebased before building on Murdock... That's why I expect minor differences that have nothing to do with the change in this PR.

**Comment by aabadie** [comment]
> just that the Murdock build of this PR will include the current master

ah yeah, I forgot that Murdock is doing a rebase on each PR... Sorry for the noise.

**Comment by miri64** [comment]
I've got another idea :-). (Don't be scared ;-))

**Comment by miri64** [comment]
Just looking at the diffs for [AVR](https://page.mi.fu-berlin.de/mlenders/atmega256rfr2-xpro.html) and [ARM7](https://page.mi.fu-berlin.de/mlenders/msba2.html) there seems to be some differences on different platforms how those new functions are compiled compared to the macros. But the behavior should remain the same.

**Comment by miri64** [comment]
Some preliminary checks at moving to `list.h` including encapsulating the more complicated functions with loops show, however, that we will gain some memory advantages with this step in the long run.

**Comment by maribu** [comment]
Interesting. I cannot reproduce the difference in `.text` for `examples/asymcute_mqttsn` for board `atmega256rfr2-xpro` when using my local toolchain. I can reproduce the 4 bytes difference with `BUILD_IN_DOCKER=1`, though. I guess it's time to update the AVR toolchain in the docker image :-)

For `tests/unittest` on the `stm32f769i-disco` board I get even 64 bytes less ROM on my toolchain and a new symbol `gnrc_pkt_append` appeared. Apparently the compiler decided to not inline that function to safe some ROM. This is a good indication that using `core/list` can result in some more ROM being saved.

**Review by maribu — APPROVED** [review]
ACK. This improves readability and type safety quite a lot. This resulted in less ROM being used with two exceptions:

- On AVR 4 bytes more ROM are used, but this is only the case due to the terribly outdated and unmaintained AVR toolchain in Ubuntu. Recent AVR toolchains generate no difference in memory consumption (either RAM or ROM)
- On `native` more ROM is used. But who care about ROM consumption on `native` :-)?

**Comment by maribu** [comment]
Let's have one recent round of Murdock before hitting merge :-)

**Pull request merged**